homeassistant/components/sonos/sensor.py

```diff
@@ -0,0 +1,161 @@
+
+from pysonos.core import SoCo
+from pysonos.exceptions import SoCoException
+
+from homeassistant.const import DEVICE_CLASS_BATTERY, PERCENTAGE, STATE_UNKNOWN
+from homeassistant.core import Event, callback
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.icon import icon_for_battery_level
+
+from . import SonosData
+from .const import DATA_SONOS, SCAN_INTERVAL, SONOS_DISCOVERY_UPDATE
+from .entity import SonosEntity
+
+_LOGGER = logging.getLogger(__name__)
+
+ATTR_BATTERY_LEVEL = "battery_level"
+ATTR_BATTERY_CHARGING = "charging"
+ATTR_BATTERY_POWERSOURCE = "power_source"
+
+
+def fetch_batery_info_or_none(soco: SoCo) -> Optional[Dict[str, Any]]:
+    """Fetch battery_info from the given SoCo object.
+
+    Returns None if the device doesn't support battery info
+    or if the device is offline.
+    """
+    with contextlib.suppress(ConnectionError, TimeoutError, SoCoException):
+        return soco.get_battery_info()
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up Sonos from a config entry."""
+
+    sonos_data = hass.data[DATA_SONOS]
+
+    async def _async_create_entity(soco: SoCo) -> Optional[SonosBatteryEntity]:
+        if not soco or soco.uid in sonos_data.battery_entities:
+            return None
+        sonos_data.battery_entities[soco.uid] = None
+        if battery_info := await hass.async_add_executor_job(
+            fetch_batery_info_or_none, soco
+        ):
+            return SonosBatteryEntity(soco, sonos_data, battery_info)
+        return None
+
+    async def _async_create_entities(event: Event):
+        if entity := await _async_create_entity(event.data.get("soco")):
+            async_add_entities([entity])
+
+    async_dispatcher_connect(hass, SONOS_DISCOVERY_UPDATE, _async_create_entities)
+
+    entities = []
+    # create any entities for devices that exist already
+    for soco in sonos_data.discovered.values():
+        if entity := await _async_create_entity(soco):
+            entities.append(entity)
+
+    if entities:
+        async_add_entities(entities)
+
+
+class SonosBatteryEntity(SonosEntity, Entity):
+    """Representation of a Sonos Battery entity."""
+
+    def __init__(self, soco: SoCo, sonos_data: SonosData, battery_info: Dict[str, Any]):
+        """Initialize a SonosBatteryEntity."""
+        super().__init__(soco, sonos_data)
+        self._battery_info = battery_info
+
+    async def async_added_to_hass(self) -> None:
+        """Register polling callback when added to hass."""
+        cancel_timer = self.hass.helpers.event.async_track_time_interval(
+            self.update, SCAN_INTERVAL
+        )
+        self.async_on_remove(cancel_timer)
+
+        self.data.battery_entities[self.unique_id] = self
+
+    async def async_seen(self, soco) -> None:
+        """Record that this player was seen right now."""
+        self._soco = soco
```

Comment:
Does the battery info also come in from the events? If so we shouldn't poll if we are getting it via events

Reply:
I'm not seeing any events triggered in my local tests when removing from its power source.

Reply:
```2021-04-21 02:18:47 DEBUG (MainThread) [pysonos.events_asyncio] Event content: <e:propertyset xmlns:e="urn:schemas-upnp-org:event-1-0"><e:property><ZoneGroupState>&lt;ZoneGroupState&gt;&lt;ZoneGroups&gt;&lt;ZoneGroup Coordinator```
...
```
MoreInfo=&quot;BattChg:NOT_CHARGING,RawBattPct:95,BattPct:100,BattTmp:29&quot;/&gt;&lt;/ZoneGroup&gt;&lt;ZoneGroup Coordinator=&quot;RINCON_48A6B828516601400&quot; 
```

I am seeing battery data in the events for the roam

Reply:
And when I plug it in

```MoreInfo=&quot;BattChg:CHARGING,RawBattPct:95,BattPct:100,BattTmp:30&quot;/&gt;&lt;/ZoneGroup&gt;&lt;ZoneGroup```

Reply:
This is provided with the `deviceProperties` subscription. Unfortunately it looks like the event payload is not as complete as calling the `/status/batterystatus` endpoint on the speaker directly. The latter includes information like the battery health and specific power source (i.e., plug vs dock).

Perhaps we can use the event payload for battery percentage, and also as a trigger to call the endpoint for more complete information when the charging status changes. Having the additional information for plug vs dock would be useful for certain automations.

Reply:
Polling when we see the charging state change to get more info is consistent with the other places we poll after we get an event that affects some state we need more information about 👍